requests-for-comments/2022-01-27-future-of-person-model.md

```diff
@@ -0,0 +1,108 @@
+
+As a result, events done between steps 1, 2 and 4 will be tracked as from person `p1` but events between steps 3 and 4 will be under person`p2`. As a result no events are lost but some will be considered as coming from a different (anonymous) person. In the above image this corresponds to the green period.
+
+This affects use cases where users access their product (Logged out) on multiple devices regularly. However, based on our analysis the impact is only on ~0.03% events, some marketing use-cases are more significantly affected.
+
+As a mitigation, if necessary, we would consider a buffer mechanism to reduce affected events to ~0.003%, we can provide further details on this solution, however we would ideally avoid implementing this as it increases the complexity of the system quite significantly.
+
+#### 2. Person properties become immutable: Changing properties applies only on new events and past data is not updated.
+
+Example: Count of active users broken down by `paying` user property each month in 2021.
+
+After this change, if `paying` property of a user in Dec 2021 is changed to `false`, count of `paying` users does not change for e.g. March 2021. The  user will be considered non-paying until their `paying` property changes again2.
+
+It also means users could be counted as both `paying=true` and `paying=false` during a given month when they change state.
+
+User properties will change over time on events, some customers may rely on the existing behavior (where the latest property is assumed for all prior events), however the new behavior is considered much more accurate by our customer feedback.
+
+Our proposed mitigation or most use-cases (e.g. events for users who have paid at some point) cohorts can be used to achieve the same results, we will document this for affected users
+
+
+### Technical implications
+
+### Benefits
+1. Faster queries: remove the need to do JOINs for any insight queries.]
+2. Unlock: Our query results no longer change unless new data has come in - we make it possible to later implement time-series caching to speed up repeated queries over long time frames (>10x improvement expected on repeated calculations).
+4. We can sample by person_id in queries, which will make it possible to build a quick way to speed up queries in huge (1Bn+ events) installations with no additional query time resource cost.
+5. Queries will require less memory and no longer OOM (making it cheaper and more reliable to operate posthog)
+3. We can shard by person_id, simplifying the sharding story a lot and removing the need for cross-shard communication when sharding.
+
+Note that some of these changes have been benchmarked [here](https://github.com/PostHog/posthog/issues/7962)
+
+
+#### Challenges
+
+6. We will need to set up some strong consistency for message ordering in plugin-server and kafka, to make sure messages are processed in the right order per user.
```

Comment:
Is this a definite necessity? This is a tough one, even if we put implementation aside.

There are many things at play here:

1. Users might want to send us old events 
2. Users may accidentally send us old events (e.g. through retries, network issues, etc)
3. Data imports/exports
4. Plugin server processing events out of order makes things easier performance-wise

#4 is mostly implementation, but the other 3 are inherent concepts that clash with this

Reply:
Agreed 100% - ingestion server changes are the meat of implementing this.

I intentionally left the details up in the air in this proposal implementation-wise and was planning to follow up with a proper plan if we decide to accept these breaking product changes. Please also take a look at these and give your thoughts on whether they're acceptable trade-offs for you.

However to answer the question:

Complexity needs to live somewhere. We're constrained by being unable to "update" data in clickhouse (even via replacing/collapsing) so this complexity naturally falls on the ingestion side of things.

I intentionally haven't defined what it means to do $set here - e.g. if you $set in an event from March 2021 what effects this should have in future events. I believe the approach we should go for is that there'll still be a set of "current" user properties (as we have currently) that will apply for _new_ events that get ingested.

Regarding ordering messages, the high-level thinking it's doable by:
- Partitioning our kafka topic by distinct_id
- Grouping messages each batch by distinct_id, enforcing ordering per distinct_id in the plugin server.

That said, I haven't put too much thought into this and out-of-order delivery. This is an area I was hoping to recruit your + @tiina303 help on if this moves on as data consistency is a key worry.

Do the higher level principles make sense? Everything said here we can iterate on should we accept the breaking changes and would like to avoid this turning into a full technical spec. :)

Reply:
I'm not as worried about a) being able to enforce order on the events we do have and b) historical exports as much as I'm worried about e.g. two people in a group are using the app at the same time, one has internet woes. Or, even with distinct IDs - events from the web and server side might arrive in a different order as they're subject to different networking conditions.

So yeah, things like that. Either way, I'm not holding back the proposal on this point, was just wondering how much this whole system relies on fully ordered ingestion. If the answer is "this system relies extremely heavily on completely in-order ingestion at all costs" then I think we need a deeper discussion, else happy to move.